### ISSUE TITLE
Truncated ASTextNode appearing in node that should not be truncated.

### ISSUE THREAD

**Issue by nickvelloff** [facebook/AsyncDisplayKit#1841]
ASCollectionNode, with ASCellNodes

Please note there is no cell re-use. They look similar (aside from background images) for testing purposes. This reproduces with _any_ string that gets truncated at smaller cell sizes.

Layout approach is pure `layoutSpecThatFits` from root `ASCellNode` down the view hierarchy.

In overriding `visibilityDidChange` I can see that the `constrainedSize` is correct for every node, including the text nodes that have the incorrect truncation. The `bounds`, however, are incorrect for the truncated text nodes.

After the resize transition, many nodes are properly laid out. When you scroll you _eventually_ hit cells that have the `ASTextNode` truncating the `NSAttributedString` to the same place as it was previous to the resize.

This happens regardless of text formatting or otherwise, and can be reproduced with a plain `NSAttributedString`.

In summary, if a `ASTextNode` has ever been truncated, it can appear again truncated with a larger `constrainedSize`, as if it was cached and reused or something. Important to note that this seems to happen for cells that are well out of any buffer ranges.

![errrr-cat](https://cloud.githubusercontent.com/assets/885369/16539373/c409b5d2-3ff5-11e6-969d-af9834ae4d11.gif)

**Comment by appleguy** [comment]
@nickvelloff Great report!  Would you be able to modify one of our examples, or make a new one, that generates this behavior?  It might be as simple as calling measure: at one size, having that display, and after a timer (or a tap) having it measure at another size.  You may be able to use the Kittens app by setting the cells to a fixed height and changing the text, as tapping the cells makes the image take up more of the cell.

I can see why this would happen for items outside of any preload ranges - but it certainly shouldn't happen at all, and indeed is probably a missing invalidateCalculatedLayout somewhere.  Should not be that hard to fix.

What framework version?  We have some changes to ASTextNode in master, although I don't think they will fix this.

**Comment by nickvelloff** [comment]
@appleguy I'm working on an example to reproduce with the lease amount of unrelated factors. I've pretty much sticking to `master` during heavy development. 
I'll post a link to the example here shortly.
Thanks for looking into this!

**Comment by nickvelloff** [comment]
@appleguy I did everything I could to boil this down to something easily reproducible. Alas, it took a lot more to build the environment necessary than I had hoped. 

The challenge is really rooted in ASDK not yet really supporting transitions from layout to layout quite yet in ASCollectionNode/View (unless I completely have missed something). Once a constrained size is received for a node, it _never asks again_. That is unless I do all the things I do to force it to transition, hence all the custom code you'll see.

My new approach to test and learn is taking advantage of Playgrounds. So I went ahead and did the example that way, in hopes of it being easier to debug. Curious as to your thoughts on it.

https://github.com/the-grid/TruncationBug-AsyncDisplayKitPlayground

**Comment by nickvelloff** [comment]
@appleguy Also you asked if this is a regression. Unfortunately I can't say, because I noticed it in a completely new app design approach, which is full stack ASDK using `ASLayoutSpec`. 
I can say however that using ASTextNode in a far less efficient manner as part of the UIKit view hierarchy did not exhibit this behavior. But that's not super useful as a comparison.

**Comment by nickvelloff** [comment]
Thought of one more thing... in testing this I did notice that even calling measure on the node at a very late `visibilityDidChange` did not cause it to correct it's bounds. I would attribute this to it already having the correct `constrainedSize`, but incorrect `bounds`. Just having dig through a lot of ASDK source I would guess it would ignore a size sent to measure that was the same as is already stored internally. Just a guess though.

**Comment by maicki** [comment]
Initial PR was reverted. Will reopen.

**Comment by knowsudhanshu** [comment]
Struggling with the similar issue. 
preload is working fine but as a new CellNode is inserted at an index out of visible range its TextNode gets clipped when I scroll directly to that indexPath ( on tap of a button ).

Have a look at attached screenshots. Please suggest a solution

![unnamed](https://user-images.githubusercontent.com/3315175/44841651-7e906e80-ac61-11e8-9880-8a09553be67e.png)


<img width="799" alt="screen shot 2018-08-30 at 2 33 34 pm" src="https://user-images.githubusercontent.com/3315175/44841907-1aba7580-ac62-11e8-8f13-96c625de3399.png">

**Comment by knowsudhanshu** [comment]
Struggling with a similar issue.
preload is working fine but as a new CellNode is inserted at an index out of visible range its TextNode gets clipped when I scroll directly to that indexPath ( on tap of a button ).

Have a look at attached screenshots. Please suggest a solution

https://user-images.githubusercontent.com/3315175/44842123-ab915100-ac62-11e8-8c36-d4168cc1ea37.png
https://user-images.githubusercontent.com/3315175/44842138-b21fc880-ac62-11e8-9172-38fcf29213bc.png
